larq_zoo/model_factory.py

```diff
@@ -0,0 +1,45 @@
+from typing import Callable, Optional, Tuple, Union
+
+import tensorflow as tf
+from zookeeper import ComponentField, Field
+from zookeeper.tf import Dataset
+
+from larq_zoo import utils
+
+# FIXME
+QuantizerType = Union[
+    tf.keras.layers.Layer, Callable[[tf.Tensor], tf.Tensor], str, None
+]
+ContraintType = Union[tf.keras.constraints.Constraint, str, None]
+DimType = Optional[int]
+
+
+class ModelFactory:
+    """A base class for Larq Zoo models. Defines some common fields."""
+
+    # Don't set any defaults here.
+    input_quantizer: QuantizerType = Field()
+    kernel_quantizer: QuantizerType = Field()
+    kernel_constraint: ContraintType = Field()
+
+    # This field is included for automatic inference of `num_clases`, if no
+    # value is otherwise provided.
+    dataset: Optional[Dataset] = ComponentField(None)
+
+    @Field
+    def num_classes(self) -> int:
+        if self.dataset is None:
+            raise TypeError("Must override either `dataset` or `num_classes`.")
```

Comment:
I find this error slightly confusing. Maybe adding something like "No dataset defined. Must override..." would help?

Reply:
Fixed in 93a3771.